File: lib/python/opentoken/src/test/opentoken/tokentransformer/decrypt_token_transformer_test.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import pickle

import pytest

from opentoken.tokentransformer.decrypt_token_transformer import DecryptTokenTransformer
from opentoken.tokentransformer.encrypt_token_transformer import EncryptTokenTransformer

class TestDecryptTokenTransformer:
    """Test cases for DecryptTokenTransformer."""

    VALID_KEY = "12345678901234567890123456789012"  # 32-character key
    INVALID_KEY = "short-key"  # Invalid short key

    def setup_method(self):
        """Set up test fixtures before each test method."""
        self.decryptor = DecryptTokenTransformer(self.VALID_KEY)
        self.encryptor = EncryptTokenTransformer(self.VALID_KEY)

    def test_serializable(self):
        """Test that DecryptTokenTransformer is serializable and deserializable."""
        decrypt_token_transformer = DecryptTokenTransformer(self.VALID_KEY)
        
        # Serialize the transformer
        serialized = pickle.dumps(decrypt_token_transformer)
        
        # Deserialize the transformer
        deserialized = pickle.loads(serialized)

        # First encrypt a token
        token = "mySecretToken"
        encrypted_token = self.encryptor.transform(token)

        # Then decrypt using deserialized transformer
        decrypted_token = deserialized.transform(encrypted_token)

        assert token == decrypted_token

    def test_constructor_valid_key_success(self):
        """Test that constructor with valid key succeeds."""
        valid_transformer = DecryptTokenTransformer(self.VALID_KEY)
        assert valid_transformer is not None

    def test_constructor_invalid_key_length_throws_value_error(self):
        """Test that constructor with invalid key length throws ValueError."""
        with pytest.raises(ValueError) as exc_info:
            DecryptTokenTransformer(self.INVALID_KEY)  # Key is too short
        
        assert "Key must be 32 characters long" == str(exc_info.value)

    def test_transform_valid_encrypted_token_returns_decrypted_token(self):
        """Test that transforming a valid encrypted token returns the decrypted token."""
        original_token = "mySecretToken"
        
        # Encrypt the token first
        encrypted_token = self.encryptor.transform(original_token)
        
        # Now decrypt it
        decrypted_token = self.decryptor.transform(encrypted_token)

        assert original_token == decrypted_token

    def test_transform_multiple_tokens_decrypts_correctly(self):
        """Test that multiple tokens can be encrypted and decrypted correctly."""
        token1 = "firstToken"
        token2 = "secondToken"
        token3 = "thirdToken"

        # Encrypt all tokens
        encrypted1 = self.encryptor.transform(token1)
        encrypted2 = self.encryptor.transform(token2)
        encrypted3 = self.encryptor.transform(token3)

        # Decrypt all tokens
        decrypted1 = self.decryptor.transform(encrypted1)
        decrypted2 = self.decryptor.transform(encrypted2)
        decrypted3 = self.decryptor.transform(encrypted3)

        # Verify all decryptions
        assert token1 == decrypted1
        assert token2 == decrypted2
        assert token3 == decrypted3

    def test_transform_same_token_encrypted_twice_both_decrypt_correctly(self):
        """Test that the same token encrypted twice produces different encrypted values but both decrypt correctly."""
        original_token = "sameToken"

        # Encrypt the same token twice (should produce different encrypted values due to random IV)
        encrypted1 = self.encryptor.transform(original_token)
        encrypted2 = self.encryptor.transform(original_token)

        # Both encrypted tokens should be different
        assert encrypted1 != encrypted2

        # But both should decrypt to the same original token
        decrypted1 = self.decryptor.transform(encrypted1)
        decrypted2 = self.decryptor.transform(encrypted2)

        assert original_token == decrypted1
        assert original_token == decrypted2

    def test_transform_special_characters_decrypts_correctly(self):
        """Test that tokens with special characters decrypt correctly."""
        special_token = "token|with|pipes|and|special!@#$%^&*()_+characters"

        encrypted = self.encryptor.transform(special_token)
        decrypted = self.decryptor.transform(encrypted)

        assert special_token == decrypted

    def test_transform_unicode_characters_decrypts_correctly(self):
        """Test that tokens with unicode characters decrypt correctly."""
        unicode_token = "token-with-unicode-你好-мир-🎉"

        encrypted = self.encryptor.transform(unicode_token)
        decrypted = self.decryptor.transform(encrypted)

        assert unicode_token == decrypted

    def test_transform_wrong_key_throws_exception(self):
        """Test that decrypting with the wrong key throws an exception."""
        original_token = "mySecretToken"
        
        # Encrypt with one key
        encrypted_token = self.encryptor.transform(original_token)
        
        # Try to decrypt with a different key
        wrong_decryptor = DecryptTokenTransformer("00000000000000000000000000000000")
        
        # Should throw an exception
        with pytest.raises(Exception):
            wrong_decryptor.transform(encrypted_token)

File: lib/java/opentoken/src/test/java/com/truveta/opentoken/tokentransformer/DecryptTokenTransformerTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.tokentransformer;

import java.security.InvalidKeyException;

import org.apache.commons.lang3.SerializationUtils;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class DecryptTokenTransformerTest {
    private DecryptTokenTransformer decryptor;
    private EncryptTokenTransformer encryptor;
    private static final String VALID_KEY = "12345678901234567890123456789012"; // 32-byte key
    private static final String INVALID_KEY = "short-key"; // Invalid short key

    @BeforeEach
    void setUp() throws Exception {
        decryptor = new DecryptTokenTransformer(VALID_KEY);
        encryptor = new EncryptTokenTransformer(VALID_KEY);
    }

    @Test
    void testSerializable() throws Exception {
        TokenTransformer decryptTokenTransformer = new DecryptTokenTransformer(VALID_KEY);
        byte[] serialized = SerializationUtils.serialize(decryptTokenTransformer);
        TokenTransformer deserialized = SerializationUtils.deserialize(serialized);

        // First encrypt a token
        String token = "mySecretToken";
        String encryptedToken = encryptor.transform(token);

        // Then decrypt using deserialized transformer
        String decryptedToken = deserialized.transform(encryptedToken);

        Assertions.assertEquals(token, decryptedToken);
    }

    @Test
    void testConstructor_ValidKey_Success() throws Exception {
        DecryptTokenTransformer validTransformer = new DecryptTokenTransformer(VALID_KEY);
        Assertions.assertNotNull(validTransformer);
    }

    @Test
    void testConstructor_InvalidKeyLength_ThrowsIllegalArgumentException() {
        Exception exception = Assertions.assertThrows(InvalidKeyException.class, () -> {
            new DecryptTokenTransformer(INVALID_KEY); // Key is too short
        });
        Assertions.assertEquals("Key must be 32 characters long", exception.getMessage());
    }

    @Test
    void testTransform_ValidEncryptedToken_ReturnsDecryptedToken() throws Exception {
        String originalToken = "mySecretToken";
        
        // Encrypt the token first
        String encryptedToken = encryptor.transform(originalToken);
        
        // Now decrypt it
        String decryptedToken = decryptor.transform(encryptedToken);

        Assertions.assertEquals(originalToken, decryptedToken);
    }

    @Test
    void testTransform_MultipleTokens_DecryptsCorrectly() throws Exception {
        String token1 = "firstToken";
        String token2 = "secondToken";
        String token3 = "thirdToken";

        // Encrypt all tokens
        String encrypted1 = encryptor.transform(token1);
        String encrypted2 = encryptor.transform(token2);
        String encrypted3 = encryptor.transform(token3);

        // Decrypt all tokens
        String decrypted1 = decryptor.transform(encrypted1);
        String decrypted2 = decryptor.transform(encrypted2);
        String decrypted3 = decryptor.transform(encrypted3);

        // Verify all decryptions
        Assertions.assertEquals(token1, decrypted1);
        Assertions.assertEquals(token2, decrypted2);
        Assertions.assertEquals(token3, decrypted3);
    }

    @Test
    void testTransform_SameTokenEncryptedTwice_BothDecryptCorrectly() throws Exception {
        String originalToken = "sameToken";

        // Encrypt the same token twice (should produce different encrypted values due to random IV)
        String encrypted1 = encryptor.transform(originalToken);
        String encrypted2 = encryptor.transform(originalToken);

        // Both encrypted tokens should be different
        Assertions.assertNotEquals(encrypted1, encrypted2);

        // But both should decrypt to the same original token
        String decrypted1 = decryptor.transform(encrypted1);
        String decrypted2 = decryptor.transform(encrypted2);

        Assertions.assertEquals(originalToken, decrypted1);
        Assertions.assertEquals(originalToken, decrypted2);
    }

    @Test
    void testTransform_SpecialCharacters_DecryptsCorrectly() throws Exception {
        String specialToken = "token|with|pipes|and|special!@#$%^&*()_+characters";

        String encrypted = encryptor.transform(specialToken);
        String decrypted = decryptor.transform(encrypted);

        Assertions.assertEquals(specialToken, decrypted);
    }

    @Test
    void testTransform_UnicodeCharacters_DecryptsCorrectly() throws Exception {
        String unicodeToken = "token-with-unicode-你好-мир-🎉";

        String encrypted = encryptor.transform(unicodeToken);
        String decrypted = decryptor.transform(encrypted);

        Assertions.assertEquals(unicodeToken, decrypted);
    }

    @Test
    void testTransform_WrongKey_ThrowsException() throws Exception {
        String originalToken = "mySecretToken";
        
        // Encrypt with one key
        String encryptedToken = encryptor.transform(originalToken);
        
        // Try to decrypt with a different key
        DecryptTokenTransformer wrongDecryptor = new DecryptTokenTransformer("00000000000000000000000000000000");
        
        // Should throw an exception
        Assertions.assertThrows(Exception.class, () -> {
            wrongDecryptor.transform(encryptedToken);
        });
    }
}
